resolwe_bio/processes/workflows/bbduk_salmon_qc.py

```diff
@@ -0,0 +1,400 @@
+                default=11,
+            )
+            fraction = FloatField(
+                label="Fraction",
+                required=False,
+                range=[0, 1.0],
+                description="Use the fraction of reads [0 - 1.0] from the orignal input file instead "
+                "of the absolute number of reads. If set, this will override the"
+                "'Number of reads' input parameter.",
+            )
+            two_pass = BooleanField(
+                label="2-pass mode",
+                default=False,
+                description="Enable two-pass mode when down-sampling. Two-pass mode is twice "
+                "as slow but with much reduced memory.",
+            )
+
+        preprocessing = GroupField(
+            Preprocessing,
+            label="Preprocessing with BBDuk",
+            hidden="!show_advanced",
+        )
+        quantification = GroupField(
+            Quantification,
+            label="Quantification (Salmon)",
+            hidden="!show_advanced",
+        )
+        downsampling = GroupField(
+            Downsampling, label="Downsampling with Seqtk", hidden="!show_advanced"
+        )
+
+    class Output:
+        """Output fields."""
+
+    def run(self, inputs, outputs):
+        """Run the workflow."""
+
+        input_bbduk = {
+            "show_advanced": True,
+            "reads": inputs.reads,
+            "min_length": inputs.preprocessing.min_length,
+            "reference": {
+                "sequences": inputs.preprocessing.adapters or [],
+                "literal_sequences": inputs.preprocessing.custom_adapter_sequences,
+            },
+            "processing": {
+                "kmer_length": inputs.preprocessing.kmer_length,
+                "hamming_distance": inputs.preprocessing.hamming_distance,
+            },
+            "operations": {
+                "k_trim": "r"
+                if inputs.preprocessing.adapters
+                or inputs.preprocessing.custom_adapter_sequences
+                else "f",
+                "min_k": inputs.preprocessing.min_k
+                if inputs.preprocessing.adapters
+                or inputs.preprocessing.custom_adapter_sequences
+                else -1,
+                "quality_trim": "r",
+                "trim_quality": inputs.preprocessing.trim_quality,
+                "quality_encoding_offset": inputs.preprocessing.quality_encoding_offset,
+                "ignore_bad_quality": inputs.preprocessing.ignore_bad_quality,
+                "maxns": inputs.preprocessing.maxns,
+            },
+        }
+
+        if inputs.reads.type.startswith("data:reads:fastq:single:"):
+            bbduk_slug = "bbduk-single"
+        elif inputs.reads.type == "data:reads:fastq:paired:":
+            input_bbduk["operations"]["trim_pairs_evenly"] = True
+            input_bbduk["operations"]["trim_by_overlap"] = True
+            bbduk_slug = "bbduk-paired"
+        else:
+            self.error("Wrong input type.")
```

Comment:
Perhaps the error message could be a bit more specific regarding the input type.